src/firebase/logger.ts

```diff
@@ -38,13 +82,139 @@ export class Logger {
     payload: Record<string, any>,
     token?: string
   ) {
-    const time = new Date().getTime();
-    const record: LogRecord = { time, severity, code, payload };
+    const now = this._nowFunc();
+    const record: LogRecord = {
+      time: now,
+      severity,
+      code,
+      payload,
+    };
     if (token) record.token = token;
 
-    // TODO: Batch multiple records and send them all at once.
-    this.logFunc({ records: [record], now: time }).then(null, err => {
-      console.log('Failed to log', record, ':', err);
+    this._updateLastActive(now);
+    this._enqueueRecords([record]);
+    this._scheduleSend();
+  }
+
+  // Sets the logger's last-active time in localStorage to now.
+  _updateLastActive(now: number) {
+    localStorage.setItem(
+      this._storagePrefix + lastActiveKeySuffix,
+      now.toString()
+    );
+  }
+
+  // Returns records from localStorage identified by the given key suffix
+  // (queuedKeySuffix or sendingKeySuffix).
+  _getRecords(keySuffix: string, prefix?: string): LogRecord[] {
+    const key = (prefix || this._storagePrefix) + keySuffix;
+    const s = localStorage.getItem(key);
+    return s ? JSON.parse(s) : [];
```

Comment:
try-catch in case the parse fails?

Reply:
done, thanks! also added a test for this